flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -114,9 +134,38 @@
      *
      * <p>The method is invoked with a set of request entries according to the buffering hints (and
      * the valid limits of the destination). The logic then needs to create and execute the request
-     * against the destination (ideally by batching together multiple request entries to increase
-     * efficiency). The logic also needs to identify individual request entries that were not
-     * persisted successfully and resubmit them using the {@code requeueFailedRequestEntry} method.
+     * asynchronously against the destination (ideally by batching together multiple request entries
+     * to increase efficiency). The logic also needs to identify individual request entries that
+     * were not persisted successfully and resubmit them using the {@code requestResult} callback.
+     *
+     * <p>From a threading perspective, the mailbox thread will call this method and initiate the
+     * asynchronous request to persist the {@code requestEntries}. NOTE: The client must support
+     * asynchronous requests and the method called to persist the records must asynchronously
```

Comment:
I am still not 100% on the concrete part handling the async calls, rather than the base handling threading. Then the concrete implementation just performing simple sync calls. For instance, in KDS sink we are using an Async Kinesis client, which is equivalent to a thread pool here and Sync Kinesis client. Please remind me of the benefit/reason of delegating async to the concrete implementation?

Reply:
As you mentioned in chat, this design is from flip-171 - idea is to have the concrete sink implementer to pull in an async client. (even though people will have the option to add a thread pool there and have a sync client if they so desire)